uniffi_macros/src/export/scaffolding.rs

```diff
@@ -0,0 +1,101 @@
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote};
+use syn::{FnArg, ItemFn, Pat, ReturnType};
+
+use super::ExportItem;
+
+pub(super) fn gen_scaffolding(item: ExportItem, mod_path: &[String]) -> syn::Result<TokenStream> {
+    match item {
+        ExportItem::Function {
+            item,
+            checksum,
+            tracked_file,
+        } => {
+            let scaffolding = gen_fn_scaffolding(&item, mod_path, checksum)?;
+            Ok(quote! {
+                #scaffolding
+                #tracked_file
+            })
+        }
+    }
+}
+
+fn gen_fn_scaffolding(
+    item: &ItemFn,
+    mod_path: &[String],
+    checksum: u16,
+) -> syn::Result<TokenStream> {
+    let name = &item.sig.ident;
+    let name_s = name.to_string();
+    let ffi_name = format_ident!("__uniffi_{}_{}_{:x}", mod_path.join("__"), name, checksum);
```

Comment:
This could potentially get out of sync with [the other code that calculates the function name](https://github.com/mozilla/uniffi-rs/blob/bb784e78399f8cb383c07e0c7a656cff7dfc26c2/uniffi_bindgen/src/interface/function.rs#L124-L129)

What do you think about:
  - Adding a `FnMetadata` method that calculates this
  -  Have the `gen_fn_scaffolding()` input `FnMetadata` rather than the `syn::ItemFn`
  - Having both piece of code call that method, so the names always stay in sync.

In general I think the scaffolding generation code and the bindings generation code should use the same types to do their work.

Also, maybe the module name isn't needed.  Right now all the functions end up in a flat namespace in the foreign bindings.

Reply:
> What do you think about:
> 
> * Adding a `FnMetadata` method that calculates this

Seems reasonable to have a shared function somewhere, I'll look into that.

> * Have the `gen_fn_scaffolding()` input `FnMetadata` rather than the `syn::ItemFn`
>
> […] 
>
> In general I think the scaffolding generation code and the bindings generation code should use the same types to do their work.

I intentionally didn't do that. I think there's going to be good reasons to write a Rust function one way or the other that don't influence the metadata / generated bindings, for example both `self: Arc<Self>` and `&self` methods are going to be methods in the bindings, but the scaffolding code probably needs to be able to tell them apart.

> Also, maybe the module name isn't needed. Right now all the functions end up in a flat namespace in the foreign bindings.

My idea was that ultimately, it would be great if (with the `nightly` feature active, or some stable solution for extracting the module path) you could generate bindings that aren't just a flat namespace. I don't think it hurts to already account for that here.